crates/fluvio/src/producer/mod.rs

```diff
@@ -29,355 +35,301 @@ pub struct TopicProducer {
+    pub(crate) _status_sender: tokio::sync::broadcast::Sender<BatchStatus>,
+    /// Used by the producer to return batch statuses to the caller.
+    pub(crate) status_receiver: Mutex<tokio::sync::broadcast::Receiver<BatchStatus>>,
+}
+
+impl Drop for ProducerInner {
+    fn drop(&mut self) {
+        debug!("Notify dispatcher shutdown");
+        self.dispatcher_shutdown.notify();
+    }
 }
 
 impl TopicProducer {
-    pub(crate) fn new(topic: String, pool: Arc<SpuPool>) -> Self {
+    pub(crate) fn new(topic: String, pool: Arc<SpuPool>, config: ProducerConfig) -> Self {
         let partitioner = Box::new(SiphashRoundRobinPartitioner::new());
+
+        let (to_dispatcher, from_dispatcher) = flume::bounded(0);
+        let (status_sender, status_receiver) = tokio::sync::broadcast::channel(100);
+        let dispatcher =
+            Dispatcher::new(pool.clone(), status_sender.clone(), from_dispatcher, config);
+        let dispatcher_shutdown = dispatcher.start();
+        let inner = Arc::new(ProducerInner {
+            dispatcher: to_dispatcher,
+            dispatcher_shutdown,
+            uid_generator: RecordUidGenerator::new(),
+            _status_sender: status_sender,
+            status_receiver: Mutex::new(status_receiver),
+        });
+
         Self {
             topic,
             pool,
             partitioner,
+            inner,
         }
     }
 
     /// Sends a key/value record to this producer's Topic.
     ///
     /// The partition that the record will be sent to is derived from the Key.
     ///
+    /// The `send` function does not wait for records to be committed to the SPU,
+    /// it simply buffers them. In order to make sure that your records have all
+    /// been actually sent, be sure to call [`TopicProducer::flush`] after sending
+    /// all the records you have.
+    ///
     /// # Example
     ///
     /// ```
     /// # use fluvio::{TopicProducer, FluvioError};
     /// # async fn example(producer: &TopicProducer) -> Result<(), FluvioError> {
     /// producer.send("Key", "Value").await?;
+    /// producer.flush().await?;
     /// # Ok(())
     /// # }
     /// ```
     #[instrument(
         skip(self, key, value),
         fields(topic = %self.topic),
     )]
-    pub async fn send<K, V>(&self, key: K, value: V) -> Result<(), FluvioError>
+    pub async fn send<K, V>(&self, key: K, value: V) -> Result<RecordUid>
     where
         K: Into<RecordKey>,
         V: Into<RecordData>,
     {
-        let record_key = key.into();
-        let record_value = value.into();
-        self.send_all(Some((record_key, record_value))).await?;
-        Ok(())
+        let key = key.into();
+        let value = value.into();
+        let record = Record::from((key, value));
+
+        {
+            // Check if there have been any failures
+            let mut receiver = self.inner.status_receiver.lock().expect("Mutex poisoned");
+            while let Ok(status) = (*receiver).try_recv() {
```

Comment:
Sure, so each time we call `producer.send()`, we want to return `Err` if something has gone wrong with a batch somewhere. The way we check that is by looking at the `status_receiver` channel, which is where the Dispatcher will send the success/failure status of each batch it flushes.

I'm using `while let Ok(status) = (*receiver).try_recv() { ... }` specifically because I don't want to _wait_ for a status in the channel, I just want to see if there is an Err _already there_. So if there's anything in the channel, it will sift through them, ignoring the successes but bubbling up `Err` if there was a `BatchStatus::Failed` or `BatchStatus::InternalError` in the channel.

Reply:
why not just use futures poll?  or use simple lock that returns (return Option<Error>) this make it harder to understand. similar to: https://github.com/infinyon/fluvio/blob/master/crates/fluvio-spu/src/control_plane/message_sink.rs